File: LittleGame-SmartestMen/Question.cpp
#include "pch.h"
#include "Question.h"
#include <format>

std::string Question::getQuestion()
{
    std::string first{ m_firstNumber < 0 ? std::format("({})",m_firstNumber):std::format("{}",m_firstNumber) };
    std::string second{ m_secondNumber < 0 ? std::format("({})",m_secondNumber) : std::format("{}",m_secondNumber) };
    std::string operate;
    switch (op)
    {
    case Question::add:
        operate = "+";
        break;
    case Question::minus:
        operate = "-";
        break;
    case Question::multiply:
        operate = "*";
        break;
    case Question::divide:
        operate = "/";
        break;
    default:
        break;
    }
    std::string result{ std::to_string(m_isAnswerCorrect ? m_correctAnswer : m_notCorrectAnswer) };
    return first + operate + second + "=" + result;
}

int Question::TickRun(int tick)
{
    rand();
    if (m_leftTime) {
        m_leftTime -= tick;
    }
    return max(m_leftTime,0);
}

bool Question::isCorrect(bool answer)
{
    if (answer == m_isAnswerCorrect) { return true; }
    else { return false; }
}

bool Question::newQuestion()
{
    int questionType;
    m_score = 1;
    if (m_difficulty < 5) {
        questionType = rand() % 2;
    }
    else {
        questionType = rand() % 4;
    }

    switch (questionType)
    {
    case 0:
        op = add;
        if (m_difficulty < 2) {
            spawnNumber(10, false);
        }
        else if (m_difficulty < 4) {
            spawnNumber(20, false);
            m_score = 2;
        }
        else {
            spawnNumber(20, true);
            m_score = 3;
        }
        break;
    case 1:
        op = minus;
        if (m_difficulty < 2) {
            spawnNumber(10, false);
        }
        else if (m_difficulty < 4) {
            spawnNumber(20, false);
            m_score = 2;
        }
        else {
            spawnNumber(20, true);
            m_score = 3;
        }
        break;
    case 2:
        op = multiply;
        if (m_difficulty < 6) {
            spawnNumber(10, false);
            m_score = 3;
        }
        else {
            spawnNumber(10, true);
            m_score = 4;
        }
        break;
    case 3:
        op = divide;
        if (m_difficulty < 6) {
            spawnNumber(10, false);
            m_score = 3;
        }
        else {
            spawnNumber(10, true);
            m_score = 4;
        }
        break;
    default:
        break;
    }
    setLeftTime();
    m_isAnswerCorrect = rand() % 2;
    switch (op)
    {
    case Question::add:
        m_correctAnswer = m_firstNumber + m_secondNumber;
        break;
    case Question::minus:
        m_correctAnswer = m_firstNumber - m_secondNumber;
        break;
    case Question::multiply:
        m_correctAnswer = m_firstNumber * m_secondNumber;
        break;
    case Question::divide:
        while (m_secondNumber == 0) { m_secondNumber += 1; }
        m_correctAnswer = m_firstNumber / m_secondNumber;
        if ((m_correctAnswer * m_secondNumber) != m_firstNumber) {
            m_isAnswerCorrect = false;
        }
        break;
    default:
        break;
    }
    if (!m_isAnswerCorrect) {
        do {
            m_notCorrectAnswer = m_correctAnswer + (rand() % 10 - 5);
        } while (m_notCorrectAnswer == m_correctAnswer);
    }
    else {
        m_notCorrectAnswer = m_correctAnswer;
    }
    return true;
}

bool Question::setDifficulty(int difficulty)
{
    m_difficulty = difficulty;
    return true;
}

bool Question::setLeftTime()
{
    switch (m_score)
    {
    case 1:m_leftTime = 1200; break;
    case 2:m_leftTime = 1400; break;
    case 3:m_leftTime = 1500; break;
    case 4:m_leftTime = 2200; break;
    case 5:m_leftTime = 2500; break;
    default:
        break;
    }
    return false;
}

bool Question::spawnNumber(int range, bool haveNegative)
{
    if (haveNegative) {
        m_firstNumber = rand() % (range * 2) - range;
        m_secondNumber = rand() % (range * 2) - range;
    }
    else {
        m_firstNumber = rand() % range;
        m_secondNumber = rand() % range;
    }
    return true;
}

File: SmartestMan-C#Test/Question.cs
using System;
using Windows.Gaming.Input.ForceFeedback;

namespace SmartestMan_C_Test
{
    public class Question
    {
        public int m_score;
        private int m_leftTime = 0;
        private int m_firstNumber;
        private int m_secondNumber;
        private int m_correctAnswer;
        private int m_notCorrectAnswer;
        private bool m_isAnswerCorrect;
        private enum operators
        {
            add,
            minus,
            multiply,
            divide
        };
        private operators op;
        private int m_result;
        private int m_difficulty = 5;

        public Question(int difficulty)
        {
            m_difficulty = difficulty;
        }

        public string getQuestion()
        {
            return $"{m_firstNumber} {op} {m_secondNumber} = ?";
        }

        public int TickRun(int tick)
        {
            m_leftTime -= tick;
            if (m_leftTime <= 0)
            {
                m_leftTime = 0;
                return 0;
            }
            return m_leftTime;
        }

        public bool isCorrect(bool answer)
        {
            return answer == m_isAnswerCorrect;
        }

        public bool newQuestion()
        {
            bool haveNegative = true;
            if (m_difficulty < 3)
            {
                haveNegative = false;
            }
            if (m_difficulty < 6)
            {
                spawnNumber(10, haveNegative);
            }
            else
            {
                spawnNumber(20, haveNegative);
            }
            setLeftTime();

            return true;
        }

        public bool setDifficulty(int difficulty)
        {
            if (difficulty < 1 || difficulty > 10)
            {
                return false;
            }
            m_difficulty = difficulty;
            return true;
        }

        private bool spawnNumber(int range, bool haveNegative)
        {
            Random rnd = new Random();
            int first = rnd.Next(0, range);
            int second = rnd.Next(0, range);
            if (haveNegative)
            {
                first = rnd.Next(-range, range);
                second = rnd.Next(-range, range);
            }
            op = (operators)rnd.Next(0, 4);
            switch (op)
            {
                case operators.add:
                    m_correctAnswer = first + second;
                    break;
                case operators.minus:
                    m_correctAnswer = first - second;
                    break;
                case operators.multiply:
                    m_correctAnswer = first * second;
                    break;
                case operators.divide:
                    if (second == 0)
                    {
                        do
                        {
                            second = rnd.Next(haveNegative ? -range : range, range);
                        } while (second == 0);
                    }
                    m_correctAnswer = first / second;
                    break;
            }
            m_firstNumber = first;
            m_secondNumber = second;
            m_isAnswerCorrect = true;
            if (rnd.Next(0, 2) == 0)
            {
                m_notCorrectAnswer = m_correctAnswer + rnd.Next(1, 10);
                m_isAnswerCorrect = false;
            }

            return true;
        }

        private bool setLeftTime()
        {
            m_leftTime = m_score switch
            {
                1 => 1200,
                2 => 1400,
                3 => 1500,
                4 => 2200,
                5 => 2500,
                _ => 3000
            };
            return true;
        }
    }

}
